src/OpenTelemetry/Trace/ActivityProcessor.cs

```diff
@@ -80,14 +82,24 @@ public void Dispose()
 
         protected virtual void Dispose(bool disposing)
         {
-            try
+            if (this.disposed)
```

Comment:
This is probably fine but I don't think necessarily it is the standard/preferred pattern. This is what I'm used to doing:

```csharp
public void Shutdown()
{
   if (!IsRunning) return;

   // Shutdown

   IsRunning = false;
}

public void Dispose()
{
   Dispose(true);
   GC.SuppressFinalize(this);
}

~ClassName()
{
   Dispose(false);
}

protected virtual void Dispose(bool disposing)
{
   Shutdown();

   if (disposing && !disposed)
   {
       // Call dispose on owned refs.

       disposed = true;
   }
}
```

The reason for that is if someone makes a mistake and forgets to call Dispose, you want the finalizer to still call Shutdown. Because it might have a thread or something it needs to signal which will prevent the process from spinning down.

Reply:
I see your point. I was concerned about running complex managed code in Finalizer thread as it might clog the entire GC (which turns an application developer's problem into a runtime library developer's problem). Thoughts?

Reply:
@reyang Also a good point. I suppose it's a tradeoff. We want users to call Dispose, and most will, negating any hit, but we need to utilize the finalizer to do the cleanup just in case the user messes up?

@noahfalk @tarekgh Would love to get an expert opinion on this if you guys have the time. Any official guidance on calling Shutdown/Close in Dispose/Finalizer?

Reply:
I think this is something we can change later, not necessarily blocking the PR.
Created #1024 for tracking.

Reply:
I didn't look at  the details, but here is the guidlines for disposing in general https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose. it is recommended to use SafeHandle instead of providing finalizer.